### ISSUE TITLE
Slack Integration

### ISSUE THREAD

**Issue by adaofeliz** [ranman/gitshots#23]
Changes:
* Local script post-commit.py to send Slack configurations: *Token* and *Channel*
* Gitshots server will add the following dependencies: *slackclient==0.15* and *websocket-client==0.32.0*

Implementation: 
* Change method: put_commit(gitshot_id)

**Comment by adaofeliz** [comment]
I will implement this if you agree.
Example: https://dl.dropboxusercontent.com/spa/y32fwqxnzbyqfvp/w2qhet9j.png

**Comment by ranman** [comment]
The only problem I see is that I wanted post-commit to be somewhat network independent. I spend most of my time on airplanes where the connectivity is bad -- would it be better to implement server-side?

**Comment by adaofeliz** [comment]
I agree and I believe that one of the most important requirements is to keep it network independent.
Probably I didn't explained it correctly, my suggestion is to add the slack configurations to the json that we push to the server (if a connection is available or latter when we try to push the ones that failed before).
In the server file app.py  we change the following method:

```python
@app.route('/api/put_commit/<ObjectId:gitshot_id>', methods=['PUT'])
def put_commit(gitshot_id):

    data = loads(request.data)
    data['ts'] = datetime.fromtimestamp(data['ts'])

    try:
        if 'where' in data:
            ts = data['where']['properties']['ts']
            data['where']['properties']['ts'] = datetime.fromtimestamp(ts)
    except:
        pass

    gitshot = mongo.db.gitshots.find_one_or_404(gitshot_id)
    gitshot.update(data)

    ret = str(mongo.db.gitshots.save(gitshot))

    # Notify git channel.
    slackclient = SlackClient(gitshot.slack.token)
    slackclient.api_call("chat.postMessage", channel=gitshot.slack.channel, text=app.config['GITSHOTS_BASE_URL'] + '/gs/' + ret)

    return ret
```

Something like that... I will analyse how to ensure that the slack tokens are not exposed to anywhere outside of the service.

**Comment by ranman** [comment]
Sounds great! I don't use slack personally but I know a lot of people do
and would appreciate this.

**Comment by adaofeliz** [comment]
It's quite cool: https://dl.dropboxusercontent.com/spa/y32fwqxnzbyqfvp/4297gpma.png
I will think about something more 'scalable' like, adding a new object to the json with 'integrations' where the gitshot server could increase the number of integrations, instead of having just slack we could have stuff like: twitter, tumblr, HipChat ... are you getting the idea?

**Comment by ranman** [comment]
That sounds good:
```json
{
...
    "extensions": {
        "tumblr": {...},
        "slack": {...}
    }
}
```

**Comment by adaofeliz** [comment]
yup! - lets do it. Tomorrow... :disappointed: ... I need to finish some other stuff. Gitshots rocks!
